### ISSUE TITLE
Querying by abstract super class with composite ID ignores discriminator values to determine subtype.

### ISSUE THREAD

**Issue by Jestar342** [nhibernate/nhibernate-core#1550]
NH 5. (though has been a problem since.. forever I think)

Self contained xUnit test case: https://github.com/Jestar342/NHibernateIssues/blob/master/NHibernateIssues/AbstractCompositeKey.cs

Result: `NHibernate.InstantiationException
Cannot instantiate abstract class or interface: NHibernateIssues.AbstractClass`

Linq Query, QueryOver and Criteria all affected the same.

If not using a composite key, the query succeeds to provide a list of instances of the subtypes.

**Comment by bahusoid** [comment]
>Querying by abstract super class with composite ID 

This part is misleading. Did you actually try to query by any derived class?
`var thing = session.CreateCriteria<DerivativeOne>().UniqueResult();`

It won't work too. The problem here is that  this mapping leads to use abstract class as identifier type. So identifier type can't be instantiated for any derived classes. 

I guess checks below for composite keys should be extended to disallow abstract classes to be used (though generating dynamic proxies for such cases would be option 2).
https://github.com/nhibernate/nhibernate-core/blob/c90b08c422f6fc17eb5f8de07fa552a6024a6b46/src/NHibernate/Cfg/XmlHbmBinding/ClassCompositeIdBinder.cs#L44-L47

You can fix this problem by wrapping you keys in some separate DTO class. Something like:
```xml
<!-- KeyPropName is a property of you AbstractClass -->
 <composite-id name="KeyPropName" > 
<!-- Key1 and Key2 are properties of ComponentKeyClass -->
  <key-many-to-one name="Key1" />
  <key-many-to-one name="Key2" />
</composite-id>
```

**Comment by Jestar342** [comment]
Yes, querying by derived class works fine - but what I am attempting to do
is query by the super class such that I can get all derivatives.

This _works_ if I encapsulate the composite key into a type.

**Comment by fredericDelaporte** [comment]
So this is indeed a "by design" issue. As explained by @bahusoid, using a non-component composite key causes the whole class to be used as identifier type. See also the [`composite-id` documentation](http://nhibernate.info/doc/nhibernate-reference/mapping.html#mapping-declaration-compositeid):

> Unfortunately, this approach to composite identifiers means that a persistent object is its own identifier. There is no convenient "handle" other than the object itself. You must instantiate an instance of the persistent class itself and populate its identifier properties before you can Load() the persistent state associated with a composite key. We will describe a much more convenient approach where the composite identifier is implemented as a separate class in [Section 7.4, “Components as composite identifiers”](http://nhibernate.info/doc/nhibernate-reference/components.html#components-compositeid).

In the case of an abstract class, this can only cause issues, since NHibernate will be unable to instantiate identifiers for a query on the base class, wrecking its loading logic.

NHibernate first instantiate the class identifier, checks if it is already in the session context for yielding the already loaded instance if it is in, and only after that, if the instance was not already in context, read the other columns. (This sentence is only a simplified explanation of the whole Loader mechanism.)

The mapping in this issue can only fail at the very first step of the NHibernate loading logic when querying by the base class.

So this case is unsupported and I do not think it is worth trying to support this, because this would go against NHibernate loader design.

Moreover, the model can be fixed in order to be supported in a relatively simple way: map the composite id as a component. This is furthermore recommended as it allows lazy-loading to work as expected, instead of always forcing immediate load of entities.

If someone wants to further check what is going on in this case, I have rewritten the test case as a [NHibernate test](https://github.com/fredericDelaporte/nhibernate-core/blob/1550/src/NHibernate.Test/NHSpecificTest/GH1550/Fixture.cs) in this [branch](https://github.com/fredericDelaporte/nhibernate-core/tree/1550) of my repository.
